Add minutes for 06-22-2023 Hardware SIG (#122)
oneapi-src/oneAPI-tab: hardware/README.rst

Meeting Notes
=============

2023-06-22
=========

Agenda
------

* Consistent timestamp reporting in Level Zero, Matias Cabral
* Level Zero port for TornadoVM + the SPIR-V code gen, Juan Fumero
* Tiles-as-devices model in Level Zero, Jaime Arteaga
* Level Zero and Unified Runtime general updates, Jaime Arteaga Molina, Alastair Murray

Attendees
---------

.. list-table::

   * - Alison L Richards, Intel
     - Jaime Arteaga Molina, Intel
     - Michal Mrozek, Intel
   * - Alastair Murray, Codeplay
     - Brice Videau, Intel
     - Matias Cabral, Intel
   * - Juan Fumero, The University of Manchester
     - Kenneth Benzie, Codeplay
     - Ben Ashbaugh, Intel
   * - Brice Goglin
     - John Daniel Holmes, Intel
     - Ruyman Reyes, Codeplay
   * - Kevin Harms, ANL
     - Xinmin Tian, Intel
     - Rod Burns, Codeplay
   * - Brandon Yates, Intel

>      - DorotheeX Marie Clotilde Balas, Intel

Reviewer: ```suggestion
     - Dorothee Marie Clotilde Balas, Intel
```

Author: thanks @alycm . Fixed.

     -

Consistent timestamp reporting in Level Zero
--------------------------------------------

* Inconsistency of metrics units for timestamps in
  Level Zero specification.
* Proposal: Explicit definition of timestamps on ns
  and resolutions in frequency.

Level Zero port for TornadoVM + the SPIR-V code gen
---------------------------------------------------

* Work on extending TornadoVM to run on oneAPI stack,
  focusing on Level Zero.

> * Current acces to heterogeneous systems is fragmented

Reviewer: ```suggestion
* Current access to heterogeneous systems is fragmented
```

Author: thanks @alycm . Fixed.

  with different stack for CPU and GPU.
* What if access to these hardware could be done from
  existing high-level programming languages.
* TornadoVM translates for instance Java Bytecode
  into CUDA, OpenCL, or SPIRV.
* Translation is done by using ACC annotations.
* TaskGraphs methods are used to identify functions
  to accelerate.
* Found that for some workloads, suggested group size
  returned by L0 driver is not as optimal as the one
  found heuristically by application.
* Garbage Collector needs to be controlled to avoid
  failures when dealing with memory moved to an accelerator.
* Some suggestions for L0:
  * Migration counters.
  * Async exception support.
  * Device aggregation.

>   * Improvements to suggested group size returned by L0.

Reviewer: ```suggestion
  * Improvements to suggested group size returned by L0.
* Question: Unified runtime would be more useful to this project as a standard?  Answer: To convince Java community to use this software stack a standard is more appealing.  Easier to justify a standard because not controlled by one party.
```

Author: thanks @alycm . Added.

Tiles-as-devices model in Level Zero
------------------------------------

* L0 adding support for new device models.
  * Mainly aimed at multi-tiled architectures
  * New environment variable: ZE_FLAT_DEVICE_HIERARCHY
    * Mode 0: Cards-as-devices model
    * Mode 1: Tiles-as-devices model (no access to root device)
    * Mode 2: Tiles-as-devices model (with access to root device)
  * Objective is to provide applications with ability to
    select best configuration for a given workload, especially
    when using architectures where access between tiles is
    more costly than local access.

>   * Changes and new APIs to be added as part of L0 v1.7.

Reviewer: ```suggestion
  * Changes and new APIs to be added as part of L0 v1.7.
* Questions:
  * How is memory allocated between the different modes?  Answer: In mode 0 get device handle and allocations split between two sub-devices (tiles), or could get sub-device and memory will be allocated on that tile.  In mode 1 get handle to a tile as device (no sub-devices) and memory is allocated on that tile.  No magic.
  * Is ZE_FLAT_DEVICE_HIERARCHY an environment variable?  Answer: Yes.
  * Specification spec will say that mode 1 in the default?  Answer: Up for debate, specification may say that default is implementation defined.
  * In Mode 2 can't do an allocation on root device?  Answer: Yes, you can.   zeDeviceGet will return all tiles.
```

Author: thanks @alycm . Added.

Level Zero and Unified Runtime general updates
----------------------------------------------

* UR Adapters have been merged to the SYCL runtime.
* License was previously MIT and now being changed to Apache
  to make it friendlier with LLVM.

> * Current UR version is v0.6 and closing to be tagging next

Reviewer: ```suggestion
* Current UR version is v0.6 and close to tagging next
```

Author: fixed.
